Common++/header/IpAddresses.h

```diff
@@ -0,0 +1,551 @@
+#ifndef PCAPPP_IP_ADDRESSES
+#define PCAPPP_IP_ADDRESSES
+
+#include <stdint.h>
+#include <string.h>
+#include <string>
+
+// TODO: remove when migration has completed
+#include "IpAddress.h"
+
+// for in_addr, in6_addr
+// TODO: remove the following includes when toInAddr() and toIn6Addr() will be removed
+#if defined(WIN32) || defined(WINx64) || defined(PCAPPP_MINGW_ENV)
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#else
+#include <netinet/in.h>
+#endif
+
+/// @file
+
+
+namespace pcpp
+{
+
+namespace experimental
+{
+	// The implementation of the classes is based on document N4771 "Working Draft, C++ Extensions for Networking"
+	// http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4771.pdf
+
+	/**
+	 * @class IPv4Address
+	 * Represents an IPv4 address (of type XXX.XXX.XXX.XXX)
+	 */
+	class IPv4Address
+	{
+	public:
+		/**
+		 * A default constructor that creates an instance of the class with unspecified/zero address
+		 */
+		IPv4Address() { memset(m_Bytes, 0, sizeof(m_Bytes)); }
+
+		/**
+		 * A constructor that creates an instance of the class out of 4-byte integer value. 
+		 * @param[in] addrAsInt The address as 4-byte integer in network byte order
+		 */
+		IPv4Address(uint32_t addrAsInt) { memcpy(m_Bytes, &addrAsInt, sizeof(m_Bytes)); }
+
+		/**
+		 * A constructor that creates an instance of the class out of 4-byte array.
+		 * @param[in] bytes The address as 4-byte array in network byte order
+		 */
+		IPv4Address(const uint8_t bytes[4]) { memcpy(m_Bytes, bytes, sizeof(m_Bytes)); }
+
+		/**
+		 * Converts the IPv4 address into a 4B integer
+		 * @return a 4B integer in network byte order representing the IPv4 address
+		 */
+		inline uint32_t toUInt() const;
+
+		/**
+		 * Returns a pointer to 4-byte array representing the IPv4 address
+		 */
+		const uint8_t* toBytes() const { return m_Bytes; }
+
+		/**
+		 * Returns a std::string representation of the address
+		 * @return A string representation of the address
+		 */
+		std::string toString() const;
+
+		/**
+		 * Determine whether the address is unspecified
+		 */
+		bool isUnspecified() const { return toUInt() == 0; }
+
+		/**
+		 * Overload of the equal-to operator
+		 */
+		bool operator==(const IPv4Address& rhs) const { return toUInt() == rhs.toUInt(); }
+
+		/**
+		 * Overload of the not-equal-to operator
+		 */
+		bool operator!=(const IPv4Address& rhs) const	{ return !(*this == rhs); }
+
+
+		// Following methods are placed for backward compatibility and will be deleted in the future
+
+		/**
+		 * Clone the object
+		 * Notice this method is deprecated and will be deleted in the future
+		 * @return A newly allocated instance which is a clone of the current instance
+		 */
+		#if __cplusplus > 201402L || _MSC_VER >= 1900
+		[[deprecated("This method is now unnecessary and added for backward compatibility. It will be deleted in the future")]]
+		#endif
+		IPv4Address* clone() const { return new IPv4Address(*this); }
```

Comment:
In current implementation this method is an abstract so the derived classes must implement it. We forced to use this method when copying an instance of `IPAddress`. However in the new implementation this method is useless because we can make a copy using an assignment operator.

In IPv4/IPv6 this method would be useful but it complicates the class interface. Besides, the user is free to use any method for creating a dynamically allocated object such as: `operator new`, `make_unique`, `make_shared`.

Reply:
yes you're right, now I remember why I added them in the first place 😃 

I think we can simply remove these methods in the new implementation

Reply:
yes you're right, now I remember why I added them in the first place 😃

I think we can simply remove these methods in the new implementation

Reply:
Done